Fixes #628:  Fix inaccurate control over concurrent thread counts

guochaosheng opened the pull request:
Fixes #628:  Fix inaccurate control over concurrent thread counts

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/alibaba/Sentinel?pullRequest=2366) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/alibaba/Sentinel?pullRequest=2366) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/alibaba/Sentinel?pullRequest=2366) it.</sub>

sczyh30 commented:
Hi, could you please illustrate your design in the PR description?

guochaosheng commented:
Describe how you did it ？
![sentinel_issues628](https://user-images.githubusercontent.com/17942861/132657535-21d26e60-a77f-4f53-856e-f16db35de4a4.png)
* Add cleanUpEffect method to TrafficShapingController.
* Updated concurrent thread limit checking logic for DefaultController. DefaultController now implements the cleanUpEffect method.
* FlowRuleChecker adds thread context variables to store checked flow control rules and adds cleanUpEffect action.
* Added checker cleanUpEffect action when FlowSlot exits

Call chain change
FlowSlot.entry -> FlowRuleChecker.checkFlow (updated) -> DefaultController.canPass (updated)
FlowSlot.exit (updated) -> FlowRuleChecker.cleanUpEffect (newly added) -> DefaultController.cleanUpEffect (newly added)
